### ISSUE TITLE
Add ECS views/filters

### ISSUE THREAD

**Issue by 2opremio** [weaveworks/scope#721]
In the same way we have views to group containers by Kubernetes pods and services we could add views to show them grouped by ECS [task](http://docs.aws.amazon.com/AmazonECS/latest/developerguide/task_defintions.html) and [service](http://docs.aws.amazon.com/AmazonECS/latest/developerguide/ecs_services.html).

In the case of ECS, the main task information is present in container labels:

```
$ docker inspect ecs-weave-ecs-demo-task-497-dataproducer-e8acc6b2e890a398b201
[...]
        "Labels": {
            "com.amazonaws.ecs.container-name": "dataproducer",
            "com.amazonaws.ecs.task-arn": "arn:aws:ecs:eu-west-1:376248598259:task/bbaa0df6-1c58-4079-8a7d-52368e8c7e7a",
            "com.amazonaws.ecs.task-definition-family": "weave-ecs-demo-task",
            "com.amazonaws.ecs.task-definition-version": "497"
        }
[...]
```

However, to obtain deeper information about the task (including the service which is supervising it), we need to access the AWS API ([there's a Go implementation](https://github.com/aws/aws-sdk-go/)), for which we need to know the cluster in which the task lives and the region of the cluster.

The region in this example is `eu-west-1` and the cluster `weave-ecs-demo-cluster`. Using the [AWS CLI](https://aws.amazon.com/cli/):

```
$ aws --region eu-west-1 ecs describe-tasks --cluster weave-ecs-demo-cluster --tasks 'arn:aws:ecs:eu-west-1:376248598259:task/bbaa0df6-1c58-4079-8a7d-52368e8c7e7a'a
{
    "failures": [], 
    "tasks": [
        {
            "taskArn": "arn:aws:ecs:eu-west-1:376248598259:task/bbaa0df6-1c58-4079-8a7d-52368e8c7e7a", 
            "overrides": {
                "containerOverrides": [
                    {
                        "name": "dataproducer"
                    }, 
                    {
                        "name": "httpserver"
                    }
                ]
            }, 
            "lastStatus": "RUNNING", 
            "containerInstanceArn": "arn:aws:ecs:eu-west-1:376248598259:container-instance/56fc000c-a466-4952-8de0-86c4d9167a01", 
            "clusterArn": "arn:aws:ecs:eu-west-1:376248598259:cluster/weave-ecs-demo-cluster", 
            "desiredStatus": "RUNNING", 
            "taskDefinitionArn": "arn:aws:ecs:eu-west-1:376248598259:task-definition/weave-ecs-demo-task:497", 
            "startedBy": "ecs-svc/9223370588559238758", 
            "containers": [
                {
                    "containerArn": "arn:aws:ecs:eu-west-1:376248598259:container/31501d39-87a1-459e-a6d0-9451f6a9c280", 
                    "taskArn": "arn:aws:ecs:eu-west-1:376248598259:task/bbaa0df6-1c58-4079-8a7d-52368e8c7e7a", 
                    "name": "dataproducer", 
                    "networkBindings": [], 
                    "lastStatus": "RUNNING", 
                    "exitCode": 0
                }, 
                {
                    "containerArn": "arn:aws:ecs:eu-west-1:376248598259:container/e35430a9-9cb5-4051-867b-b7c559056b78", 
                    "taskArn": "arn:aws:ecs:eu-west-1:376248598259:task/bbaa0df6-1c58-4079-8a7d-52368e8c7e7a", 
                    "name": "httpserver", 
                    "networkBindings": [
                        {
                            "protocol": "tcp", 
                            "bindIP": "0.0.0.0", 
                            "containerPort": 80, 
                            "hostPort": 80
                        }
                    ], 
                    "lastStatus": "RUNNING", 
                    "exitCode": 0
                }
            ]
        }
    ]
}
```

It seems like the task doesn't have service information. `ecs-svc/9223370588559238758` indicates that the task was started by a service but not exactly which one. I don't know what's `9223370588559238758` but it's not the service identifier:

```
$ aws --region eu-west-1 ecs list-services --cluster weave-ecs-demo-cluster 
{
    "serviceArns": [
        "arn:aws:ecs:eu-west-1:376248598259:service/weave-ecs-demo-service"
    ]
}
```

So we will have to maintain a reverse lookup table for task->service (or container->service).

```
$ aws --region eu-west-1 ecs describe-services --cluster weave-ecs-demo-cluster --services arn:aws:ecs:eu-west-1:376248598259:service/weave-ecs-demo-service
{
    "services": [
        {
            "status": "ACTIVE", 
            "taskDefinition": "arn:aws:ecs:eu-west-1:376248598259:task-definition/weave-ecs-demo-task:497", 
[...]
```

The same applies for task->cluster , since the cluster is needed in first instance to obtain any further information to what's provided in the labels:

```
$ aws --region eu-west-1 ecs list-clusters
{
    "clusterArns": [
        "arn:aws:ecs:eu-west-1:376248598259:cluster/default", 
        "arn:aws:ecs:eu-west-1:376248598259:cluster/weave-ecs-demo-cluster", 
        "arn:aws:ecs:eu-west-1:376248598259:cluster/micro-wiki-demo"
    ]
}
```

```
$ aws --region eu-west-1 ecs list-tasks --cluster weave-ecs-demo-cluster
{
    "taskArns": [
        "arn:aws:ecs:eu-west-1:376248598259:task/673e70ed-49e5-41c2-98f1-acf9d43a6d56", 
        "arn:aws:ecs:eu-west-1:376248598259:task/6cb5cb6e-6344-42dc-bdf4-66b3894b87f2", 
        "arn:aws:ecs:eu-west-1:376248598259:task/bbaa0df6-1c58-4079-8a7d-52368e8c7e7a"
    ]
}
```

In order to obtain the region, we can query the [internal AWS metadata service](http://docs.aws.amazon.com/AWSEC2/latest/UserGuide/ec2-instance-metadata.html#instancedata-data-retrieval):

```
$ curl -s http://169.254.169.254/latest/dynamic/instance-identity/document | jq -r .region
eu-west-1
```

**Comment by ekimekim** [comment]
this is a little problematic because it requires the user's AWS credentials - I suppose we could make this info gathering contingent on an arg or env var or the like, but it's an annoying extra step.

**Comment by 2opremio** [comment]
> this is a little problematic because it requires the user's AWS credentials

Not from inside the instances.

**Comment by 2opremio** [comment]
The instances will need to have the correct IAM rights. But I believe that properly configured ECS machines should give us all we need: http://docs.aws.amazon.com/AmazonECS/latest/developerguide/instance_IAM_role.html (note that we need to perform a superset of the actions done by the ECS agent, which will be running in the same machines as the probes)

**Comment by 2opremio** [comment]
I stand corrected. We will need permissions for some of the `Describe*` and `List*` actions: http://docs.aws.amazon.com/AmazonECS/latest/developerguide/ecs-supported-iam-actions-resources.html

**Comment by ekimekim** [comment]
Looks like the cluster name *is* present in the container labels:
```
$ docker inspect ecs-weave-ecs-demo-task-1-dataproducer-d6969be68587f4cfa101 | jq '.[].Config.Labels'
{
  "com.amazonaws.ecs.cluster": "weave-ecs-demo-cluster",
  "com.amazonaws.ecs.container-name": "dataproducer",
  "com.amazonaws.ecs.task-arn": "arn:aws:ecs:us-east-1:664268111851:task/a16963d0-d67f-4a51-8626-cd5afb89ad63",
  "com.amazonaws.ecs.task-definition-family": "weave-ecs-demo-task",
  "com.amazonaws.ecs.task-definition-version": "1"
}
```
maybe that's a recent change?

**Comment by 2opremio** [comment]
Could be, my analysis is a bit old.

**Comment by ekimekim** [comment]
Looks like the `ecs-svc/9...` thing is a deployment id:
```
            "deployments": [
                {
                    "taskDefinition": "arn:aws:ecs:us-east-1:664268111851:task-definition/weave-ecs-demo-task:1",
                    "status": "PRIMARY",
                    "id": "ecs-svc/9223370558208240205",
                    "updatedAt": 1478646535.602,
                    "runningCount": 3,
                    "pendingCount": 0,
                    "createdAt": 1478646535.602,
                    "desiredCount": 3
                }
            ],
```

**Comment by ekimekim** [comment]
ok, so based on my own playing around and reading docs, this is my understanding:

* tasks map to a task definition
* services map to a current task definition, and also more generally to a list of deployments
* one task definition can be used by many services
* deployments map to a task definition
* tasks map to the deployment that caused them to be started
* deployments only exist until all tasks that they started are stopped

So the bad news is that we can't unambiguously map from a service to a task definition (and therefore to a task). The good news is we found something better! We can get a deployment from the task, then find the service that contains that deployment. This lets us identify the service of a task even if multiple services are using identical task definitions, or if a service is currently in the middle of a rolling upgrade.

So the full plan is:
1. get cluster and task arn from labels
2. get region from metadata
3. list-services on cluster
4. describe-service for all services to get deployment -> service reverse map
5. describe-task using task arn to get started by, which is a deployment id
6. look up service for task using deployment -> service map

**Comment by 2opremio** [comment]
Deployments are new to me and I cannot find any documentation on them. Where did that JSON come from?

**Comment by ekimekim** [comment]
describe-service

**Comment by ekimekim** [comment]
if it helps, here's some go docs on what a "Deployment" entails: https://docs.aws.amazon.com/sdk-for-go/api/service/ecs/#Deployment

**Comment by ekimekim** [comment]
```
    // The tag specified when a task is started. If the task is started by an Amazon
    // ECS service, then the startedBy parameter contains the deployment ID of the
    // service that starts it.
    StartedBy *string `locationName:"startedBy" type:"string"`
```
so it seems that not all tasks map to deployments, but that's fine.

**Comment by 2opremio** [comment]
> so it seems that not all tasks map to deployments, but that's fine.
Yep, you can run tasks without services (just like you can run pods without replication controllers in K8s)

**Comment by 2opremio** [comment]
Fixed by #2026
